### PR TITLE
mgr/diskprediction_local: Support Python 3.9 for disk prediction module.

### PR DESCRIPTION
**Pull request by kalaspuffar**
mgr/diskprediction_local: Support Python 3.9 for disk prediction module.

I've been using the disk prediction module to keep track of the health data of my drives and monitor it regularly. Lately, they are showing as STALE and I realized the module was no longer available. Probably because the module requires Python 3.7 or earlier to run and the default version of Python for Debian Bullseye is 3.9. 

Debian is usually really careful when they upgrade to newer versions so for them to go to 3.9 and keep a stable branch there seems like a good indicator that upgrading the module to 3.9 too would be beneficial to some users.

This PR aims to update some of the module dependencies and change the required version of Python. I've run the unit tests and they pass for this code which is promising. I'm trying to test it in my test cluster and I'll return with more on that later.

I'm creating the PR here for discussion and to see if I perhaps have misunderstood the situation or missed some implementation detail.

Thank you for your time and keep up the great work.

Fixes: https://tracker.ceph.com/issues/50196
Signed-off-by: Daniel Persson <mailto.woden@gmail.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/common/options/CMakeLists.txt [hunk 1]
```diff
@@ -74,13 +74,6 @@ if(FREEBSD)
 endif()
 
 set(mgr_disabled_modules "")
-if(WITH_MGR)
-  # https://tracker.ceph.com/issues/45147
-  if(Python3_VERSION VERSION_GREATER_EQUAL 3.8)
```
Comment (tchaikov) [inline src/common/options/CMakeLists.txt:79]: my major concern is https://tracker.ceph.com/issues/45147 . if python3.9 is proved to be safe, we should instead make python3.8 an exception, something like
```cmake
if(Python3_VERSION VERSION_EQUAL 3.8)
```

Comment (kalaspuffar) [reply src/common/options/CMakeLists.txt:79]: Hi @tchaikov 

That sounds like a reasonable precaution. To extend that thought, there is no guarantee it will work in 3.7 either. The new libraries I've updated to are recommended from 3.9 and up.

They might work just fine for 3.7- but not guaranteed.

So I see two approaches.

Either create a test harness that tests all three versions to guarantee interoperability and test 3.8 manually to see if we can reproduce the issue.

Or if we just want to exclude the version, that could also be ok when some of the distributors of supported distros made the jump over 3.8 anyway.

Best regards
Daniel

Comment (tchaikov) [reply src/common/options/CMakeLists.txt:79]: > Hi @tchaikov

hi Daniel,

> 
> That sounds like a reasonable precaution. To extend that thought, there is no guarantee it will work in 3.7 either. The new libraries I've updated to are recommended from 3.9 and up.
> 
> They might work just fine for 3.7- but not guaranteed.
> 

agreed. we do disable [0] this mgr module for ubuntu focal, which comes with python 3.8. and it is enabled for centos 8 stream, which comes with python3.6.8.

---
[0] search "mgr module disabled for" in https://jenkins.ceph.com/job/ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=focal,DIST=focal,MACHINE_SIZE=gigantic/49953//consoleFull

> So I see two approaches.
> 
> Either create a test harness that tests all three versions to guarantee interoperability and test 3.8 manually to see if we can reproduce the issue.

a little more background first:

- the integration test environment we have: ubuntu focal and centos 8
- the existing test harness: `qa/tasks/mgr/test_module_selftest.py`

we only have python 3.6.8 and python3.8 for integration tests. yeah, we *could* have python3.7 as well using tox driven tests, as debian buster comes with python3.7 and Luminous (Ceph 12.2). but the question is, what distro we should test for? are we test on python3.7 just for the sake of completeness? if that's the case, i'm not sure if it's worthy doing so. and i think we should focus on current/future stable releases. they are more likely to ship python3.8 and newer.

regarding reproduce on 3.8,  please see https://tracker.ceph.com/issues/45147 , i'd avoid repeat what we did 2 years ago unless we have something new. for instance, if python had a fix of https://bugs.python.org/issue15751 for python3.8.

> 
> Or if we just want to exclude the version, that could also be ok when some of the maintainers of supported distros made the jump over 3.8 anyway.

debian stable comes with python3.9. 3.8 is much more popular than 3.7
> 
> Best regards Daniel

Comment (kalaspuffar) [reply src/common/options/CMakeLists.txt:79]: Hi @tchaikov 

I've read the resources you supplied, and I've not dived deep into testing on different versions or similar, but most of the reports are talking about issues with NumPy. This PR will change the NumPy version to a newer one that should support newer versions of Python. Then again, to be safe, I see no reason why we should not exclude 3.8.

I've made a small commit changing this PR to follow your advice. 

Thank you for reviewing.

Best regards
Daniel

File: qa/tasks/mgr/test_module_selftest.py [hunk 2]
```diff
@@ -52,12 +52,6 @@ def test_influx(self):
 
     def test_diskprediction_local(self):
         self._load_module("selftest")
-        python_version = self.mgr_cluster.mon_manager.raw_cluster_cmd(
-            "mgr", "self-test", "python-version")
-        if tuple(int(v) for v in python_version.split('.')) >= (3, 8):
```
Comment (tchaikov) [inline qa/tasks/mgr/test_module_selftest.py:57]: if the solution is to *skip* 3.8, probably instead of removing the whole `if` block, we should s/>=/==/.

File: src/pybind/mgr/diskprediction_local/requirements.txt [hunk 3]
```diff
@@ -1,3 +1,3 @@
-numpy==1.15.1
-scipy==1.1.0
-scikit-learn==0.19.2
+numpy==1.23.0
+scipy==1.8.1
+scikit-learn==1.1.1
```
Comment (tchaikov) [inline src/pybind/mgr/diskprediction_local/requirements.txt:3]: these changes should go to a dedicated commit. also, i am not convinced that we need to bump up these dependencies. they are pin'ed to these versions because they are "supported" versions. unless the major distros we are supposed to support package, for instance python3-numpy v1.23.0, we should bump up the numpy's version.

File: src/pybind/mgr/tox.ini [hunk 4]
```diff
@@ -1,7 +1,7 @@
 [tox]
 envlist =
     py3,
-    py37,
+    py39,
```
Comment (tchaikov) [inline src/pybind/mgr/tox.ini:4]: ditto. justifications needed. currently we build packages for CentOS 8, openSUSE 15.3 and Ubuntu focal, and test the built packages on CentOS 8 and Ubuntu focal. the tox based test is the first guard for the suspicious changes. it is supposed to test the typical settings and fail the tests which would otherwise break the tests in our integration tests exercising the packages. 

probably we should bump the version to 3.9 in near future. but probably not now.

File: src/pybind/mgr/diskprediction_local/requirements.txt (hunk 3 above)
Comment (kalaspuffar) [reply src/pybind/mgr/diskprediction_local/requirements.txt]: Hi @tchaikov 

I understand that bumping versions could be a larger endeavor, but I hoped to get some updates to these libraries because there have been a bunch of good changes. Some of them are security-related other performance. But let's leave that for another person to create a commit when they change in supported distros.

Best regards
Daniel

**Comment by kalaspuffar** [comment]
Hi @tchaikov 

I've removed the unnecessary changes. Now, we only check if the Python version is 3.8 and skip to running the disk prediction code on these installations.

Best regards
Daniel

**Comment by tchaikov** [comment]
@kalaspuffar hi Daniel, could you squash the commits into a single one?

**Comment by kalaspuffar** [comment]
> @kalaspuffar hi Daniel, could you squash the commits into a single one?

Fixed.

**Review by tchaikov — APPROVED** [review]

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
https://pulpito.ceph.com/kchai-2022-08-23_13:19:39-rados-wip-kefu-testing-2022-08-22-2243-distro-default-smithi/

failures tracked by

- https://tracker.ceph.com/issues/57256
- https://tracker.ceph.com/issues/57163
- https://tracker.ceph.com/issues/55789

**Comment by spiderbeg** [comment]
@kalaspuffar Can the disk prediction module run on Python 3.9?
